Add template plugin and renaming script

goosemania opened the pull request:
closes #2859
https://pulp.plan.io/issues/2859

jortel commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,55 @@
+    #    unique_together = ()
+
+    @classmethod
+    def natural_key_fields(cls):
```
This seems dangerous since _unique_together_ can be either: ("name", "version")` or `(("name", "version"), ..)`

goosemania replied:
Do you suggest to complete this task first? https://pulp.plan.io/issues/3025

jortel replied:
I'm not sure it needs to be done first but suggest you remove this impl here for this PR.

jortel replied:
No, 3025 has a checklist item to vet on pulp-dev.  Suggest you just leave this in for now so you are not blocked.

jortel commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,55 @@
+
+    # path = models.TextField(blank=False, null=False)
+    # digest = models.TextField(blank=False, null=False)
+    # objects = 
```
I think a comment like:
```
# DEFINE FIELDS HERE
```
instead of commented out fields definitions would be clearer.

jortel commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,55 @@
+    TYPE = 'plugin-template'
+
+    # path = models.TextField(blank=False, null=False)
+    # digest = models.TextField(blank=False, null=False)
```
The default is _blank=False_ and _null=False_ so just adds clutter here.

jortel commented:
Did you consider instead replacing all forms of _plugin-template_, we defined a set of replaceable tokens like: 

- `<NAME>` - The name.
- `<NAME-CAMEL>` - The name formatted to camel case.
- `<NAME-UNDERSCORE>` - The name formatted with underscores.
- `<NAME-DASH>` - The name formatted with dashes.

Then replaced the tokens with the formatted name.  This approach seems clearer to the plugin writer as to what needs to be replaced.  And, more straight forward for the script to do the substitution.

Examples:

```
class <NAME-CAMEL>Content(Content):
    """
    The <NAME> content type.
    """
```

```
class <NAME-CAMEL>Importer(Importer)
    """
    An importer for <NAME>Content content. 
    """

    TYPE = <NAME-UNDERSCORE>
```

```
class <NAME-CAMEL>Publisher(Publisher)
    """
    A publisher for <NAME>Content content. 
    """

    TYPE = <NAME-UNDERSCORE>
```

```
class <NAME-CAMEL>ContentViewSet(ContentViewSet):
    endpoint_name = '<NAME-DASH>'
```

goosemania commented:
@jortel , that's a good point.
The way it is done now the code works even without renaming. I'm not sure there is a lot of value in it though. And following your suggestion, maybe it make sense to use some template engine, like Jinja. Or it's too much for this task? :)

jortel commented:
Hard to say about the template engine.  I haven't used any of them so not sure what would be more/less effort.

asmacdo commented on setup.py:
```diff
@@ -0,0 +1,24 @@
+]
+
+setup(
+    name='pulp-plugin-template',
```
one thing that could confuse a newbie is that this is a functioning setup.py. One benefit of switching to jinja would be that it would be clear that this file is a template for the plugin rather than the installer for the template.

asmacdo commented on rename.py:
```diff
@@ -0,0 +1,105 @@
+                  os.path.join(root_dir, pulp_plugin_name))
+
+    # replace text
+    for dir_path, dirs, files in os.walk(root_dir):
```
If we use a copy/rename workflow, it would be good to ignore hidden files that start with "."

goosemania replied:
For now the assumption is that plugin writer doesn't do anything (except maybe just checking if template kinda works right away) before renaming, so only git-related stuff is hidden. It is ignored explicitly. Sounds ok?

goosemania commented on rename.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    Check if specified name is compliant with requirements for it
+    """
+    return bool(re.match(r'^[a-z][0-9a-z_]{2,15}$', name))
```
Will do.
There are few reasons why I suggest to put at least some limit. One limit for us is a directory name - depends on filesystem but usually max filename length is 255. We also use plugin name in class names so for readability and to fit line length :) it makes sense to ask for something relatively short I think. And 16 characters seem reasonable to me and it can be easily changed by anyone in case they need more.

asmacdo replied:
Seems reasonable. I think it is helpful to add a comment though, or else future pulp engineers might assume there is a  correctness reason that it is limited to 16.

asmacdo commented on pulp_plugin_template/app/viewsets.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class PluginTemplatePublisherViewSet(platform.PublisherViewSet):
+    endpoint_name = 'plugin-example'
```
s/plugin-example/plugin-template/

goosemania replied:
good catch :)

goosemania commented:
Addressed most of the feedback, added docs and docstrings.

I'm on the fence on choosing between:
 - "working template even without renaming" aka class names and other objects contain `PluginTemplate` substring, simply replace by that substring
 - "explicit template string like <NAME-CAMEL>", simply replace by that substring
 - use template engine like Jinja

Currently the first option is implemented. Let me know if someone have strong opinion towards other options or have any new ideas.

asmacdo commented on README.rst:
```diff
@@ -0,0 +1,115 @@
+
+``$ http POST http://localhost:8000/api/v3/repositories/foo/publishers/plugin-template/bar/distributions/ some=params``
+
+Sync repository ``foo`` using importer ``bar``
```
s/importer/Importer/

asmacdo commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,86 @@
+    It is responsible for publishing metadata and artifacts
+    which belongs to a specific repository.
+
+    For more details see Plugin Writer's Guide.
```
I don't think we need this in the class level docstrings. I think it makes the most sense early in the documentation, [maybe here](https://github.com/pulp/plugin_template/pull/1/files#diff-88b99bb28683bd5b7e3a204826ead112R17). I also like that it is included in the module level docstring.

asmacdo commented on pulp_plugin_template/app/serializers.py:
```diff
@@ -0,0 +1,61 @@
+    field3 = serializers.CharField()
+
+    class Meta:
+        fields = platform.ContentSerializer.Meta.fields + ('field1', 'field2', 'field3')
```
Good idea to include this example, this would be easy to overlook.

asmacdo commented on rename.py:
```diff
@@ -0,0 +1,117 @@
+    for dir_path, dirs, files in os.walk(dst_root_dir):
+
+        # skip directories which don't need any text replacement
+        for dir in set(IGNORE_DIRS).intersection(dirs):
```
I think we can get rid of this part, and the IGNORE_DIRS contant. Both __pycache__ and .git are in IGNORE_COPYTREE, so I don't think these sets will ever intersect.

goosemania replied:
Good catch, it was needed if we did renaming in-place without copy the whole directory

dkliban commented on README.rst:
```diff
@@ -0,0 +1,118 @@
+``./rename.py your_plugin_name``
+ 
+Check `Plugin Writer's Guide <http://docs.pulpproject.org/en/3.0/nightly/plugins/plugin-writer/index.html>`__ 
+for more details ans suggestions on plugin implementaion.
```
s/ans/and/

dkliban commented on README.rst:
```diff
@@ -0,0 +1,118 @@
+Check `Plugin Writer's Guide <http://docs.pulpproject.org/en/3.0/nightly/plugins/plugin-writer/index.html>`__ 
+for more details ans suggestions on plugin implementaion.
+
+Below are some ideas how to document your plugin.
```
s/ideas how/ideas for how/

dkliban commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,82 @@
+    which belongs to a specific repository.
+    """
+    TYPE = 'plugin-template'
+
```
We should mention that any fields specific to this publisher can also be specified here.

dkliban commented on pulp_plugin_template/app/models.py:
```diff
@@ -0,0 +1,82 @@
+    downloading of the content and saving it to Pulp.
+    """
+    TYPE = 'plugin-template'
+
```
Same as with the publisher, the importer for this plugin can have some additional fields.
